sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/spanner/SpannerIOWriteTest.java

```diff
@@ -263,6 +263,17 @@ private void verifyBatches(Iterable<Mutation>... batches) {
 
   @Test
   public void noBatching() throws Exception {
+
+    // This test uses a different mock/fake because it explicitly does not want to populate the
+    // Spanner schema.
+    FakeServiceFactory fakeServiceFactory = new FakeServiceFactory();
+    ReadOnlyTransaction tx = mock(ReadOnlyTransaction.class);
+    when(fakeServiceFactory.mockDatabaseClient().readOnlyTransaction()).thenReturn(tx);
+
+    // Capture batches sent to writeAtLeastOnce.
+    when(fakeServiceFactory.mockDatabaseClient().writeAtLeastOnce(mutationBatchesCaptor.capture()))
+        .thenReturn(null);
+
```

Comment:
This doesn't seem to be used. It seems we should either:
- Pass `fakeServiceFactory` to `withServiceFactory` and the verify calls below, or
- Just remove `fakeServiceFactory` and use `serviceFactory` here instead. It seems to be capable of verifying that the schema isn't read, so I'm not sure there's value in building another fake that doesn't have a schema.

I'd prefer the latter, but I'd be fine with the former if you think that's better.

Reply:
Actually on second thought I wonder if we should just leave this test completely unchanged (i.e. call `verifyBatches`, don't check that the DB schema is never read). The fact the schema isn't read seems like an implementation detail that we don't need to explicitly verify. What do you think?

Reply:
This test specifically tests that the pipeline stage for reading the schema (required for grouping/batching) are not being used, verifying that the batching is being bypassed. Yes, this is an implementation detail, but I added this part of the test so that future changes do not break this behaviour.

I cannot use the already-created serviceFactory here because that does not give me access to the mock ReadOnlyTransaction that I need to verify that executeQuery() is never called.

Reply:
Ah I see. But should you be using `fakeServiceFactory` throughout this test then? There are still references to `serviceFactory`

Reply:
Oops, yes, correct. Fixed now.

> Ah I see. But should you be using `fakeServiceFactory` throughout this test then? There are still references to `serviceFactory`

Oops, yes, correct! Fixed now.